// @mui/x-date-pickers/legacy/internals/hooks/usePicker/usePicker.js
import { usePickerValue } from './usePickerValue';
import { usePickerViews } from './usePickerViews';
import { usePickerLayoutProps } from './usePickerLayoutProps';
export var usePicker = function usePicker(_ref) {
  var props = _ref.props,
    valueManager = _ref.valueManager,
    wrapperVariant = _ref.wrapperVariant,
    inputRef = _ref.inputRef,
    additionalViewProps = _ref.additionalViewProps,
    validator = _ref.validator,
    autoFocusView = _ref.autoFocusView;
  var pickerValueResponse = usePickerValue({
    props: props,
    valueManager: valueManager,
    wrapperVariant: wrapperVariant,
    validator: validator
  });
  var pickerViewsResponse = usePickerViews({
    props: props,
    inputRef: inputRef,
    additionalViewProps: additionalViewProps,
    autoFocusView: autoFocusView,
    propsFromPickerValue: pickerValueResponse.viewProps
  });
  var pickerLayoutResponse = usePickerLayoutProps({
    props: props,
    wrapperVariant: wrapperVariant,
    propsFromPickerValue: pickerValueResponse.layoutProps,
    propsFromPickerViews: pickerViewsResponse.layoutProps
  });
  return {
    // Picker value
    open: pickerValueResponse.open,
    actions: pickerValueResponse.actions,
    fieldProps: pickerValueResponse.fieldProps,
    // Picker views
    renderCurrentView: pickerViewsResponse.renderCurrentView,
    hasUIView: pickerViewsResponse.hasUIView,
    shouldRestoreFocus: pickerViewsResponse.shouldRestoreFocus,
    // Picker layout
    layoutProps: pickerLayoutResponse.layoutProps
  };
};
